### PR TITLE
Bugfix for passing None args to user defined Triton kernel

### PR DESCRIPTION
**Pull request by SamGinzburg**
This is a PR to fix the following issue: https://github.com/pytorch/pytorch/issues/115344

In short, passing None as an arg. to a Triton kernel would cause problems:

Short repro of the specific bug fixed here:
```
import triton
import triton.language as tl

@triton.autotune( # E: Untyped decorator makes function "sin_kernel" untyped  [misc]
    configs=[
        triton.Config({'BLOCK_SIZE': 32}, num_stages=5, num_warps=2),
        triton.Config({'BLOCK_SIZE': 64}, num_stages=4, num_warps=4),
    ],
    key=['n_elements']
)
@triton.jit # E: Untyped decorator makes function "sin_kernel" untyped  [misc]
def sin_kernel( # E: Function is missing a return type annotation  [no-untyped-def]
    in_ptr0,
    out_ptr,
    n_elements,
    BLOCK_SIZE: "tl.constexpr",
):
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    if in_ptr0 is not None:
        x = tl.load(in_ptr0 + offsets, mask=mask)
    else:
        x = 0.
    output = tl.sin(x)
    tl.store(out_ptr + offsets, output, mask=mask)

import torch

def sin_triton(x, out):
    n_elements = out.numel()
    sin_kernel[(n_elements,)](x, out, n_elements)

x = torch.randn(65, device="cuda")
out = torch.empty_like(x)
out_compiled = torch.empty_like(x)

sin_triton_compiled = torch.compile(fullgraph=True)(sin_triton)

for first in (x, None):
    sin_triton(first, out)
    sin_triton_compiled(first, out_compiled)
    torch.testing.assert_close(out, out_compiled)
```
I've added a unit test to catch this issue in the future and the tests in "test/inductor/test_triton_kernels.py"

topic: not user facing

================================================================================
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138260

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D64615061](https://our.internmc.facebook.com/intern/diff/D64615061)

### CODE REVIEW

File: test/inductor/test_triton_kernels.py [hunk 1]
```diff
@@ -2070,6 +2070,47 @@ def fn(x):
...
+            sin_kernel[(n_elements,)](x, out, n_elements)
+
+        x = torch.randn(65, device="cuda")
+        out = torch.empty_like(x)
+        out_compiled = torch.empty_like(x)
+        sin_triton_compiled = torch.compile(fullgraph=True)(sin_triton)
+        for first in (x, None):
+            sin_triton(first, out)
+            sin_triton_compiled(first, out_compiled)
+            torch.testing.assert_close(out, out_compiled)
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:2112]: Let's use `self.assertEqual` instead, as the outputs should be identical.

File: test/inductor/test_triton_kernels.py [hunk 2]
```diff
@@ -2070,6 +2070,47 @@ def fn(x):
...
+        # https://github.com/pytorch/pytorch/issues/115344
+        @triton.autotune(  # E: Untyped decorator makes function "sin_kernel" untyped  [misc]
+            configs=[
+                triton.Config({"BLOCK_SIZE": 32}, num_stages=5, num_warps=2),
+                triton.Config({"BLOCK_SIZE": 64}, num_stages=4, num_warps=4),
+            ],
+            key=["n_elements"],
+        )
+        @triton.jit  # E: Untyped decorator makes function "sin_kernel" untyped  [misc]
+        def sin_kernel(  # E: Function is missing a return type annotation  [no-untyped-def]
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:2084]: I think, we don't need those `# E: ...`s.

File: test/inductor/test_triton_kernels.py [hunk 3]
```diff
@@ -2070,6 +2070,47 @@ def fn(x):
...
+
+        def sin_triton(x, out):
+            n_elements = out.numel()
+            sin_kernel[(n_elements,)](x, out, n_elements)
+
+        x = torch.randn(65, device="cuda")
+        out = torch.empty_like(x)
+        out_compiled = torch.empty_like(x)
+        sin_triton_compiled = torch.compile(fullgraph=True)(sin_triton)
+        for first in (x, None):
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:2109]: Would it make sense to unroll the loop, so that if the test fails we know whether it was with `x` or `None`? Also, I'd suggest doing `None` first, as it's more likely to have issues after `None` than before, I believe.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -5321,6 +5317,22 @@ def codegen(self, wrapper):
 
         # Call to kernel
         self.codegen_comment(wrapper)
+
+        """
+        Filter out None args.
+
+        see _precompile_config(...) and https://github.com/pytorch/pytorch/issues/115344
```
Comment (aakhundov) [inline torch/_inductor/ir.py:5324]: If we want to refer to `_precompile_config` here, let's mention it's in the `triton_heuristics.py`?

File: torch/_inductor/runtime/triton_heuristics.py [hunk 5]
```diff
@@ -481,12 +481,41 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
...
+        We also don't want to modify self.fn.
+
+        We know that we removed something from the signature if:
+            1. It's in compile_meta["constants"]
+            2. It isn't a constant we already know about
+                Note: The value of interest has already been added to compile_meta['constants'],
+                    so we use self.fn.constexprs instead.
+            3. It isn't in the compile_meta signature
+        """
+        self.arg_names = self.fn.arg_names
```
Comment (aakhundov) [inline torch/_inductor/runtime/triton_heuristics.py:498]: Why do we need this? It doesn't look to be changed anywhere, so should we keep using the `self.fn.arg_names` as before? If we really want a copy in `self.arg_names`, let's do `list(self.fn.arg_names)` to shallow-copy, so that we don't keep a reference (assuming `self.fn.arg_names` is a `list`, can lead to unexpected side effects down the road).

File: torch/_inductor/runtime/triton_heuristics.py [hunk 6]
```diff
@@ -481,12 +481,41 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
...
+        removed_from_sig = set(compile_meta["constants"].keys())
```
Comment (aakhundov) [inline torch/_inductor/runtime/triton_heuristics.py:499]: Wondering if we could have a better name for `removed_from_sig`? Otherwise a bit hard to understand what this holds (and why it's usable for both `call_args` and `def_args` below).

File: torch/_inductor/runtime/triton_heuristics.py [hunk 7]
```diff
@@ -481,12 +481,41 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
...
+        known_constants = {
+            arg for i, arg in enumerate(self.arg_names) if i in self.fn.constexprs
+        }
+        removed_from_sig = removed_from_sig.difference(known_constants)
+        removed_from_sig = removed_from_sig.difference(
+            set(compile_meta["signature"].keys())
```
Comment (aakhundov) [inline torch/_inductor/runtime/triton_heuristics.py:505]: Can `compile_meta["signature"].keys()` intersect with `compile_meta["constants"].keys()` (used as the base for `removed_from_sig`? From the code in the `wrapper.py` seems disjoint.

File: torch/_inductor/runtime/triton_heuristics.py [hunk 8]
```diff
@@ -481,12 +481,41 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
...
+        )
+
         call_args = [
             arg
-            for i, arg in enumerate(self.fn.arg_names)
-            if i not in self.fn.constexprs
+            for i, arg in enumerate(self.arg_names)
+            if i not in self.fn.constexprs and arg not in removed_from_sig
```
Comment (aakhundov) [inline torch/_inductor/runtime/triton_heuristics.py:511]: One thing I don't quite understand is why we can't use`compile_meta["constants"].keys()` instead of `removed_from_sig` here and below? Constants get baked in during `triton.compile`, so should neither be in `def_args` nor `call_args`? And our `None` arg is in the `compile_meta["constants"].keys()` by now?

**Review by aakhundov — COMMENTED** [review]
Thanks, @SamGinzburg! Looks good overall. Added some minor comments and questions.

**Comment by aakhundov** [comment]
cc @zou3519 as this fixes our long-standing `None` arg issue :)

File: torch/_inductor/runtime/triton_heuristics.py (hunk 5 above)
Comment (SamGinzburg) [reply torch/_inductor/runtime/triton_heuristics.py:498]: it turned out that this is a bug, so we don't need this! (we can just use self.fn.arg_names)

File: torch/_inductor/runtime/triton_heuristics.py (hunk 8 above)
Comment (SamGinzburg) [reply torch/_inductor/runtime/triton_heuristics.py:511]: So the logic here is this:

Elsewhere (in `UserDefinedTritonKernel.codegen(...)`) we filter out the None arg from `raw_args` which was added in `define_user_defined_triton_kernel(...)`.

So now we have to generate the proper bindings in Python so that our Triton kernel can be called---we have to filter out the parameter that we removed from the function signature (i.e., `call_args` and `def_args`). Specifically, we remove `None` args that aren't marked as `tl.constexpr`.

`compile_meta["constants"].keys()` does have the constants that were baked in earlier (including the None arg which was added in `define_user_defined_triton_kernel(...)`). However, this results in the following checks in `call_args` and `def_args` not being strict enough (and breaking other tests).

For example:

In the test `KernelTests.test_triton_kernel_equal_to_1_arg_dynamic_False`, we can see the following intermediate state:
```python
(Pdb) compile_meta["constants"].keys()
dict_keys(['BLOCK_SIZE', 'half_n_elements'])
(Pdb) known_constants # Note: this is just self.fn.constexprs mapped to the arg names
{'BLOCK_SIZE'}
(Pdb) compile_meta["signature"].keys()
dict_keys(['in_ptr0', 'in_ptr1', 'out_ptr', 'half_n_elements'])
(Pdb)
```

If we just use `compile_meta["constants"].keys()` as a stand-in for `removed_from_sig` in the following code:
```python
call_args = [
    arg
    for i, arg in enumerate(self.fn.arg_names)
    if i not in self.fn.constexprs and arg not in removed_from_sig
]
```

Then this would result in `half_n_elements` being removed (breaking the test). This specific failure happens because equal_to_1 args are treated as constants in `define_user_defined_triton_kernel`---but not included in `self.fn.constexprs`. These args are in the function signature and we don't want to filter these out from `call_args` or `def_args`.

`removed_from_sig = compile_meta["constants"].keys() - compile_meta["signature"].keys()` is also insufficient, breaking many tests. This is because `removed_from_sig` can end up containing `tl.constexpr` if not explicitly filtered out.

Example: `KernelTests.test_triton_kernel_with_imported_symbol_with_custom_name`:

These are the correct values for `call_args` and `def_args`---we see that `BLOCK_SIZE` (i.e., `tl.constexpr`) is included in def_args:
```python
(Pdb) call_args
['in_ptr', 'out_ptr', 'n_elements']
(Pdb) def_args
['in_ptr', 'out_ptr', 'n_elements', 'BLOCK_SIZE']
```
If we don't filter out the `tl.constexpr` values then this breaks the test.


Lastly, `removed_from_sig = compile_meta["constants"].keys() - tl.constexpr` is also insufficient, because tests such as `CustomOpTests.test_autotune_unbacked`---also because of equal_to_1 args (which are in the signature and constants---but we don't want to mark as having been removed).


So in conclusion, the logic we want is that we want to find the None args that we just removed---which we can find by:
1. Take `compile_meta["constants"].keys()`---which contains the value we want
2. Filtering out known `tl.constexpr` (since this doesn't match 1, due to equal_to_1 args)
3. Filtering out args present in the function signature

`removed_from_sig = compile_meta["constants"].keys() - tl.constexpr - compile_meta["signature"].keys()`

Comment (SamGinzburg) [reply torch/_inductor/runtime/triton_heuristics.py:511]: So in short here what we want to do here is identify which None arguments we just removed from the list---and make sure we don't include them in `call_args` or `def_args`.

There are three sources of information at play:
1. `compile_meta["constants"].keys()`
2. `self.fn.constexprs`
3. `compile_meta["signature"].keys()`

Using (1) alone is insufficient because it isn't a strict enough measure of which argument we just removed---for example equal_to_1 args get included here, but not as a `tl.constexpr`, so this breaks several tests.

Using (1) - (3) is not sufficient because the set difference could include constexprs, which can be present in (1).

Using (1) - (2) breaks things because of equal_to_1 args, since we aren't filtering those out either (just the `None` args). 

The combination of (1) - (2) - (3) ensures that we only include the exact values we filtered out in `removed_from_sig`---so that the logic in `call_args` and `def_args` is sound (i.e., this set tracks the exact values we filtered out earlier).

File: torch/_inductor/runtime/triton_heuristics.py (hunk 7 above)
Comment (SamGinzburg) [reply torch/_inductor/runtime/triton_heuristics.py:505]: I think I address this here: https://github.com/pytorch/pytorch/pull/138260#discussion_r1806760766

File: torch/_inductor/runtime/triton_heuristics.py (hunk 6 above)
Comment (SamGinzburg) [reply torch/_inductor/runtime/triton_heuristics.py:499]: maybe something like `filtered_none_args`?

File: torch/_inductor/ir.py [hunk 9]
```diff
@@ -5321,6 +5323,22 @@ def codegen(self, wrapper):
...
+
+        Two cases for a None arg:
+        1. The arg is already tl.constexpr, so leave it in
+        2. The arg is not tl.constexpr so we have to remove it
+        """
+        constexpr_args_set = set(constexpr_indices)
+        raw_args = [
+            arg
+            for idx, arg in enumerate(raw_args)
+            if (arg is not None) or (arg is None and idx in constexpr_args_set)
```
[the remark this answers is not in the document]
Comment (SamGinzburg) [reply torch/_inductor/ir.py:5340]: `raw_args` can contain constexpr args. The logic here is:

1. We want to include any argument that isn't `None` (lhs)
2. If the arg is `None`, check if it is a constexpr---if so leave it in (rhs)

The non-`None` constexprs from raw_args should continue to be included here because they would be captured by the lhs of the statement (```arg is not None```).

File: torch/_inductor/runtime/triton_heuristics.py (hunk 6 above)
Comment (aakhundov) [reply torch/_inductor/runtime/triton_heuristics.py]: Yeah, this is more explicit, thanks!

File: torch/_inductor/runtime/triton_heuristics.py (hunk 8 above)
Comment (aakhundov) [reply torch/_inductor/runtime/triton_heuristics.py]: Thanks for the detailed explanation, this makes sense.

**Comment by SamGinzburg** [comment]
@SamGinzburg has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64615061).

**Review by aakhundov — APPROVED** [review]

**Comment by SamGinzburg** [comment]
@SamGinzburg has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64615061).
